### ISSUE TITLE
4.6.1 not able to replace node 

### ISSUE THREAD

**Issue by slivne** [scylladb/scylladb#10337]
replace falls with 

```
storage_service - replace[7b2becdd-7917-4c17-8eb6-a0db06c2c8ee]: Failed to send heartbeat: std::runtime_error (node_ops_cmd_check: Node 192.168.41.225 rejected node_ops_cmd=8 from node=192.168.41.225 with ops_uuid=7b2becdd-7917-4c17-8eb6-a0db06c2c8ee, pending_node_ops={}, the node ops is unknown)
```

scylla.yaml

```
cluster_name: *** Cluster
num_tokens: 256
data_file_directories:
     - /var/lib/cassandra/data/vol
commitlog_directory: /var/lib/cassandra/commitlog
saved_caches_directory: /var/lib/cassandra/saved_caches
commitlog_sync: periodic
commitlog_sync_period_in_ms: 10000
commitlog_segment_size_in_mb: 32
seed_provider:
                    - class_name: org.apache.cassandra.locator.SimpleSeedProvider
      parameters:
                              - seeds:  "audiencedb001.spotx.den01.pop,audiencedb004.spotx.den01.pop,audiencedb007.spotx.den01.pop"
listen_address:  192.168.41.225
broadcast_address: 192.168.41.225
listen_on_broadcast_address: false
native_transport_port: 9042
native_shard_aware_transport_port: 19042
read_request_timeout_in_ms: 5000
write_request_timeout_in_ms: 2000
cas_contention_timeout_in_ms: 1000
endpoint_snitch: GossipingPropertyFileSnitch
rpc_address:  192.168.41.225
rpc_port: 9160
api_port: 10000
api_address: 127.0.0.1
batch_size_warn_threshold_in_kb: 64
authenticator: PasswordAuthenticator
authorizer: CassandraAuthorizer
initial_token:
broadcast_rpc_address: 192.168.41.225
hints_directory: /var/lib/cassandra/hints
view_hints_directory: /var/lib/cassandra/view_hints
hinted_handoff_enabled: true
max_hint_window_in_ms: 10800000 
permissions_validity_in_ms: 300000
permissions_update_interval_in_ms: 100000
permissions_cache_max_entries: 10000
partitioner: org.apache.cassandra.dht.Murmur3Partitioner
commitlog_total_space_in_mb:  635625
storage_port: 7000
ssl_storage_port: 7001
start_native_transport: true
start_rpc: false
incremental_backups: false
snapshot_before_compaction: false
auto_snapshot: false
tombstone_warn_threshold: 1000
tombstone_failure_threshold: 100000
murmur3_partitioner_ignore_msb_bits: 12
replace_address: 192.168.41.225
enable_repair_based_node_ops: false
failure_detector_timeout_in_ms: 1200000
```

```
Datacenter: realtime
====================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address         Load       Tokens       Owns    Host ID                               Rack
UN  192.168.42.136  1.83 TB    256          ?       be8748be-69be-4ca2-9ed2-319b0c0f2222  rack1
UN  192.168.42.137  1.81 TB    256          ?       9ea4a90b-432a-447a-8e7d-d7641dedd4bb  rack1
UN  192.168.42.138  1.83 TB    256          ?       2541ead4-da71-4c4c-848f-d759e49faf8a  rack1
UN  192.168.42.139  1.8 TB     256          ?       0c1143ff-4367-48c9-9fe2-802d1f4c8a10  rack1
UN  192.168.42.140  1.87 TB    256          ?       59073f2b-d9a0-44fa-ae80-93e9695a1783  rack1
UN  192.168.42.141  1.91 TB    256          ?       021556b1-5d32-4dd7-857a-0e98e8678184  rack1
UN  192.168.42.142  1.93 TB    256          ?       40d3c919-cfb3-4f5f-8e82-31bc7bbe8bf4  rack1
UN  192.168.42.143  1.95 TB    256          ?       abb8cb4c-4386-493d-b1a8-7196dbd37dad  rack1
UN  192.168.42.128  1.96 TB    256          ?       52f4b176-5b07-42e7-86f2-9f9ffeb93fa3  rack1
UN  192.168.42.129  1.84 TB    256          ?       e39f14db-afa3-4b49-9a5f-020950a736bf  rack1
UN  192.168.42.130  1.87 TB    256          ?       dd0b738d-2f8e-4a4d-9ca5-0b8450dff445  rack1
UN  192.168.42.131  1.72 TB    256          ?       916aa22f-a394-4f84-8acb-ffaa0d321309  rack1
UN  192.168.42.133  1.86 TB    256          ?       b1bd9f6d-b932-49cb-ae54-9f725ada2705  rack1
UN  192.168.42.134  1.88 TB    256          ?       0806a400-93a4-4a79-b6c6-20d3ebb61f8e  rack1
UN  192.168.42.135  1.85 TB    256          ?       346633c9-af91-46c1-9f20-8f08fd7cf502  rack1
UN  192.168.42.144  1.9 TB     256          ?       10413591-232f-4388-99c4-d15443d0c15c  rack1
UN  192.168.42.145  1.83 TB    256          ?       eea0661f-25d2-4cd9-a7e2-0b5dda5e1e4a  rack1
UN  192.168.42.146  1.85 TB    256          ?       b873fa4d-8469-49bc-9b51-c1c775b80aa2  rack1
UN  192.168.42.147  1.85 TB    256          ?       fee46e85-5586-48c0-9ea4-65e1480d7ecd  rack1
UN  192.168.42.148  1.86 TB    256          ?       c104e310-3d3a-456b-8ba1-6a0e7152629f  rack1
UN  192.168.42.149  1.8 TB     256          ?       a8af8c58-6cb2-4918-8978-087a588c2c86  rack1
UN  192.168.42.150  1.81 TB    256          ?       28038226-3fdc-4597-a5ed-058e6df66bbf  rack1
UN  192.168.42.151  1.78 TB    256          ?       33768cdc-8208-4351-ab9f-eaea4507e3ee  rack1
UN  192.168.41.232  1.84 TB    256          ?       c29114ab-782f-433d-877e-7ecba979c1ad  rack1
UN  192.168.41.233  1.9 TB     256          ?       ffa53dfd-f848-45af-ab6a-45e236d372e5  rack1
UN  192.168.41.234  1.93 TB    256          ?       c411b4b8-5eb3-4ca1-9932-e78304a65bb7  rack1
UN  192.168.41.235  1.84 TB    256          ?       bf1b0eea-9a1f-4108-82e8-e829c2569f38  rack1
DN  192.168.41.225  1.87 TB    256          ?       2a10d145-569c-4a37-b5ff-a6cc1ea177d7  rack1
UN  192.168.41.226  1.91 TB    256          ?       67a2b3cd-a71d-4bb4-9e0b-478c60c9f206  rack1
UN  192.168.41.228  1.81 TB    256          ?       02f7bcdb-f5f2-45e8-b9fe-e67c163cfbe8  rack1
UN  192.168.41.229  1.8 TB     256          ?       d8ee34e3-6aa7-4837-b6ae-2103c130ce99  rack1
UN  192.168.41.230  1.74 TB    256          ?       b4d79371-aa1b-4a4d-86c6-84a8c23a1434  rack1
UN  192.168.42.120  1.82 TB    256          ?       3191e1a6-5b93-454e-bbc7-cb35af77bf48  rack1
UN  192.168.42.121  1.91 TB    256          ?       c8ba37cd-1ad1-449d-8970-985d22ee1890  rack1
UN  192.168.42.122  1.84 TB    256          ?       fd0e86c0-810e-4d3c-9561-5ae212327905  rack1
UN  192.168.42.123  1.85 TB    256          ?       288c915f-3d00-49d7-a10b-b1da15a54304  rack1
UN  192.168.42.124  1.79 TB    256          ?       a88c32df-5369-40ea-8474-81937e795dde  rack1
UN  192.168.42.125  1.9 TB     256          ?       34ee4b71-8401-4efc-86b4-e8f33ef07224  rack1
UN  192.168.42.126  1.9 TB     256          ?       bac85a18-fd00-4b22-a7d0-d9899e2d894d  rack1
UN  192.168.42.127  1.99 TB    256          ?       c0e25f76-f9d9-4947-a59e-1a6f02101345  rack1
UN  192.168.42.116  1.8 TB     256          ?       7a684440-6022-4ca3-a4be-2a3651d5384f  rack1
UN  192.168.42.117  1.84 TB    256          ?       bbcc910a-e205-41bb-adff-df88c388d95a  rack1
UN  192.168.42.118  1.82 TB    256          ?       e4dc86d2-5cbc-42dc-bb59-6bf53c27d67f  rack1
UN  192.168.42.119  1.81 TB    256          ?       0158341c-6548-4bc9-a307-e30e04cab117  rack1
UN  192.168.41.201  1.91 TB    256          ?       ec2ecbe5-9378-4c96-9807-74276fe9fb0a  rack1
UN  192.168.41.202  1.91 TB    256          ?       956fa92f-f4f6-4e55-98b3-b237b31204ae  rack1
UN  192.168.41.203  1.84 TB    256          ?       48d13090-1164-492e-80e7-1046d5727545  rack1
UN  192.168.41.204  1.86 TB    256          ?       55c4c5d3-5959-4890-8d3d-10440d34c857  rack1
UN  192.168.41.206  1.87 TB    256          ?       a81882e3-3cf1-4c91-8201-f976edaef68d  rack1
UN  192.168.41.207  1.73 TB    256          ?       6a7cd283-0673-43fa-bb0f-c6a74b2dd3a7  rack1
UN  192.168.41.196  1.72 TB    256          ?       47184799-d324-43ad-a3ce-936f790e2f94  rack1
UN  192.168.41.198  1.85 TB    256          ?       12715393-dd60-4e68-9357-56e2d7ab6e02  rack1
UN  192.168.41.199  1.84 TB    256          ?       79fc455a-5628-4b8e-a4d7-b11485f4a2b4  rack1
```

scylla is started with
```
/usr/bin/scylla --log-to-syslog 1 --log-to-stdout 0 --default-log-level info --lock-memory 1 --network-stack posix --io-properties-file=/etc/scylla.d/io_properties.yaml --cpuset 1-15,17-31
```

**Comment by slivne** [comment]
The user was not able to replace a node with and without repair-based-operations

Nicholas Peshek can you please add the logs of the bootstrapping node

**Comment by asias** [comment]
Yes, please provide both the full log of the replacing node and two other existing nodes in the cluster.

**Comment by npeshek-spotx** [comment]
OK, I worked through this: Since a node's out, we're seeing sustained CPU usage across the cluster when a node crashes and shard 0 cannot respond to node operations in a timely manner. Due to the hardcoded 30 second watchdog timer on node operation metadata
https://github.com/scylladb/scylla/blob/branch-4.6/service/storage_service.hh#L125
the uuid gets ejected from the pending list on the coordinator node and the repair fails when it attempts to move to the heartbeat operations. Is there anyway we could make that value configurable for emergency situations?

**Comment by asias** [comment]
Hello, 30s should be more than enough. Which scylla version are you using? I think it might be caused by a bug which could cause deadlock in some cases. The bug is fixed in recent release. 

```
commit c18cc5dc680cc3b8294c5cf204967fcbf1078a4b
Author: Pavel Emelyanov <xemul@scylladb.com>
Date:   Thu Feb 10 17:18:45 2022 +0300

    log: Fix silencer to be shard-local and logger-global
    
    There are two issues with the log silencer.
    
    First, it shuts up logging on all shards, while the intention was to
    avoid reentrances into syslog by one shard. This is because the _level
    field is shared between shards and setting it to "silent" is visible
    everywhere.
    
    Second, it doesn't fully solve the original problem. Lowering the level
    on a logger will prevent it from self-reentrance, but it doesn't keep
    different loggers from nesting into each other. Instead of silencing
    a logger it should stop the whole shard from logging.
    
    Said that, the fix is in introducing the thread-local "be quiet" bit
    that would make all loggers stop logging, but only on a specific shard.
    This new bit is +1 condition, but it sits outside of the builtin-expect
    and gets into the slow-path of the logging code.
    
    fixes: #1012
    tests: unit(dev), ioinfo(dev)
    
    Signed-off-by: Pavel Emelyanov <xemul@scylladb.com>
    Message-Id: <20220210141845.25085-1-xemul@scylladb.com>
```

**Comment by npeshek-spotx** [comment]
I was on Scylladb 4.6.2. I have since upgraded the cluster to a custom compiled version with the timeout extended and I can get replaces to start finally. The other "symptom" I see is it takes a long time for gossip to "settle" on restart, on the range of several thousand seconds, and both of these behaviors seem to have started when we upgraded the cluster to the 4.6 series.

**Comment by slivne** [comment]
@npeshek-spotx - 4.6.3 is out with the fix for (https://github.com/scylladb/scylla/issues/10337#issuecomment-1108099788) - can you try and see if that resolves your issue.

@asias 4.6. does have https://github.com/scylladb/scylla/issues/8488 - yet its not supposed to cause an issue / @npeshek-spotx gossip also has its own scheduling group - under advanced dashboard of monitoring you should be able to see if gossip is actually "using" cpu on shard 0 and how much - can you please check that out (please specify which scylla version custom/official you are using so its clear).

**Comment by asias** [comment]
> I was on Scylladb 4.6.2. I have since upgraded the cluster to a custom compiled version with the timeout extended and I can get replaces to start finally. The other "symptom" I see is it takes a long time for gossip to "settle" on restart, on the range of several thousand seconds, and both of these behaviors seem to have started when we upgraded the cluster to the 4.6 series.

It is not supposed to have such long gossip settle time. Please try 4.6.3 to see if helps.

**Comment by pgoron** [comment]
> OK, I worked through this: Since a node's out, we're seeing sustained CPU usage across the cluster and shard 0 cannot always respond to node operations in a timely manner. Due to the hardcoded 30 second watchdog timer on node operation metadata https://github.com/scylladb/scylla/blob/branch-4.6/service/storage_service.hh#L125 the uuid gets ejected from the pending list on the coordinator node and the repair fails when it attempts to move to the heartbeat operations. Is there anyway we could make that value configurable for emergency situations?

We observe a similar issue on a cluster with 70 nodes with scylla 4.6.3 preventing new nodes to join cluster. It takes 1m30 to send bootstrap_prepare message to all nodes in cluster. watchdog ejects node_ops before heartbeat has a chance to start.

**Comment by asias** [comment]
> > OK, I worked through this: Since a node's out, we're seeing sustained CPU usage across the cluster and shard 0 cannot always respond to node operations in a timely manner. Due to the hardcoded 30 second watchdog timer on node operation metadata https://github.com/scylladb/scylla/blob/branch-4.6/service/storage_service.hh#L125 the uuid gets ejected from the pending list on the coordinator node and the repair fails when it attempts to move to the heartbeat operations. Is there anyway we could make that value configurable for emergency situations?
> 
> We observe a similar issue on a cluster with 70 nodes with scylla 4.6.3 preventing new nodes to join cluster. It takes 1m30 to send bootstrap_prepare message to all nodes in cluster. watchdog ejects node_ops before heartbeat has a chance to start.

Hi @pgoron . Thanks for reporting. 

The bootstrap_prepare messages are sent to the 70 nodes in parallel. It should not took 1m30s to finish such small messages. 
Probably, some of the nodes are extremely slow for some reason.  So the total time to finish the bootstrap_prepare took 1m30 long.  Was the system super busy at the time?

**Comment by pgoron** [comment]
Hi @asias , attaching debug logs of storage_service of shard 0 of bootstrapping node, between sending of bootstrap_prepare message and start of heartbeat_updater:
[timing.log](https://github.com/scylladb/scylla/files/8904102/timing.log)

It seems to me that a large part of the time is spent on "Updating pending ranges for keyspace=system_distributed_everywhere"

For the record, at time of the log, cluster size was 67 nodes with 256 tokens/node.

**Comment by asias** [comment]
> Hi @asias , attaching debug logs of storage_service of shard 0 of bootstrapping node, between sending of bootstrap_prepare message and start of heartbeat_updater: [timing.log](https://github.com/scylladb/scylla/files/8904102/timing.log)
> 
> It seems to me that a large part of the time is spent on "Updating pending ranges for keyspace=system_distributed_everywhere"
> 
> For the record, at time of the log, cluster size was 67 nodes with 256 tokens/node.

Thanks. I will try to reproduce.

**Comment by asias** [comment]
I created: https://github.com/scylladb/scylla/issues/10817

**Comment by pgoron** [comment]
Thanks @asias. Any plan to backport it to 4.6.4 ?

I've also same symptom than reported by @npeshek-spotx above:
>  The other "symptom" I see is it takes a long time for gossip to "settle" on restart

On our side, gossip take 25min to settle when a restarting an node on a 70-nodes cluster. Is such duration expected ?
Could it be related ? should we open another issue ? if yes what investigation material do you need ?
Thanks again for your help.

**Comment by asias** [comment]
> Thanks @asias. Any plan to backport it to 4.6.4 ?

We need some time to propagate to master and backport. 

> I've also same symptom than reported by @npeshek-spotx above:
> 
> > The other "symptom" I see is it takes a long time for gossip to "settle" on restart
> 
> On our side, gossip takes 25min to settle when restarting an node on a 70-nodes cluster. Is such duration expected ? Could it be related ? should we open another issue ? if yes what investigation material do you need ? Thanks again for your help.

The long settle thing happens only on the new bootstrapping node? Do you see that if you restart one of the existing node. You can enable the `--logger-log-level gossip=trace` to get some debug info.

**Comment by npeshek-spotx** [comment]
Finally able to get back to this ticket. We experience the same issue with the long settle on every restart. For my cluster we’re usually seeing between 1500 and 3000 seconds waiting for the gossip to “settle”

when I turn on gossip tracing, I get a massive amount of cross-node chatter with cache heat balancing. I’ll get you full logs from a node tomorrow.

Other things of note: compared to my functional 4.5 clusters, shard 0 seems to be the vulnerable shard more often than not and I see timeouts accessing the system tables frequently (system.local and system.peers at a minimum seem slower than they should be). Not sure if everything fits together, but I wanted to update with where I’m at. I managed to monkey patch enough to get nodes joining, but the problems remain.

**Comment by asias** [comment]
Hello @npeshek-spotx and @pgoron I sent another [PR](https://github.com/scylladb/scylla/pull/10892) to fix the long gossip settle issue. 

Hopefully, along with https://github.com/scylladb/scylla/pull/10850, they will solve both issues for you.

**Comment by pgoron** [comment]
Nice @asias ! we are eager to test both fixes once they reach 4.6 branch. Thanks for your help.

**Comment by geobeau** [comment]
We had a crash of the whole cluster at the end of a replace operation:

Here is the log of a node that crashed (new node is `10.228.191.13`):
```
July 19th 2022, 23:47:01.448  0x40aa931
July 19th 2022, 23:47:01.448 Aborting on shard 0.
July 19th 2022, 23:47:01.448  0x299842c
July 19th 2022, 23:47:01.448  /opt/scylladb/libreloc/libstdc++.so.6+0xad306
July 19th 2022, 23:47:01.448  /opt/scylladb/libreloc/libc.so.6+0x268a3
July 19th 2022, 23:47:01.448  0x29dafba
July 19th 2022, 23:47:01.448  /opt/scylladb/libreloc/libc.so.6+0x3d2a1
July 19th 2022, 23:47:01.448  0x7fd78d8e7a1f
July 19th 2022, 23:47:01.448 terminate called after throwing an instance of 'std::runtime_ ERROR'
July 19th 2022, 23:47:01.448  0x29a0050
July 19th 2022, 23:47:01.448  0x3e27018
July 19th 2022, 23:47:01.448  what():  Host 10.228.191.13 does not have HOST_ID application_state
July 19th 2022, 23:47:01.448  0x3e57e72
July 19th 2022, 23:47:01.448  /opt/scylladb/libreloc/libstdc++.so.6+0xad29b
July 19th 2022, 23:47:01.448  0xf4e6fa
July 19th 2022, 23:47:01.448  /opt/scylladb/libreloc/libstdc++.so.6+0xa1a45
July 19th 2022, 23:47:01.448 Backtrace:
July 19th 2022, 23:46:44.077 INFO  2022-07-19 23:46:44,077 [shard 0] storage_service - replace[49272acc-b8f9-4431-97c0-a629c58722e5]: Marked ops done from coordinator=10.228.191.13
... (elapsed to beginning of bootstrap, grep on "10.228.191.13")
July 19th 2022, 23:46:44.077 INFO  2022-07-19 23:46:44,077 [shard 0] storage_service - replace[49272acc-b8f9-4431-97c0-a629c58722e5]: Marked ops done from coordinator=10.228.191.13
July 19th 2022, 11:07:57.203 WARN  2022-07-19 11:07:57,203 [shard 0] gossip - Node 10.228.191.13 does not contain SUPPORTED_FEATURES in gossip or system table
July 19th 2022, 11:06:10.367 INFO  2022-07-19 11:06:10,366 [shard 0] token_metadata - Added node 10.228.191.13 as pending replacing endpoint which replaces existing node 10.228.198.13
July 19th 2022, 11:06:10.367 INFO  2022-07-19 11:06:10,366 [shard 0] storage_service - replace[49272acc-b8f9-4431-97c0-a629c58722e5]: Added replacing_node=10.228.191.13 to replace existing_node=10.228.198.13, coordinator=10.228.191.13
July 19th 2022, 11:05:42.534 WARN  2022-07-19 11:05:42,534 [shard 0] gossip - Node 10.228.191.13 does not contain SUPPORTED_FEATURES in gossip or system table
July 19th 2022, 11:04:51.655 WARN  2022-07-19 11:04:51,654 [shard 0] gossip - Node 10.228.191.13 does not contain SUPPORTED_FEATURES in gossip or system table
July 19th 2022, 11:04:51.516 INFO  2022-07-19 11:04:51,515 [shard 0] gossip - InetAddress 10.228.191.13 is now UP, status = UNKNOWN
July 19th 2022, 11:04:51.328 INFO  2022-07-19 11:04:51,328 [shard 0] gossip - 60000 ms elapsed, 10.228.191.13 gossip quarantine over
July 19th 2022, 11:03:50.991 INFO  2022-07-19 11:03:50,991 [shard 0] gossip - FatClient 10.228.191.13 has been silent for 30000ms, removing from gossip
July 19th 2022, 11:03:20.108 WARN  2022-07-19 11:03:20,108 [shard 0] gossip - Node 10.228.191.13 does not contain SUPPORTED_FEATURES in gossip or system table
July 19th 2022, 11:03:20.041 INFO  2022-07-19 11:03:20,040 [shard 0] gossip - InetAddress 10.228.191.13 is now UP, status = UNKNOWN
July 19th 2022, 10:56:18.434 INFO  2022-07-19 10:56:18,433 [shard 0] gossip - 60000 ms elapsed, 10.228.191.13 gossip quarantine over
July 19th 2022, 10:55:18.124 INFO  2022-07-19 10:55:18,124 [shard 0] gossip - FatClient 10.228.191.13 has been silent for 30000ms, removing from gossip
July 19th 2022, 10:53:56.328 WARN  2022-07-19 10:53:56,328 [shard 2] gossip - failure_detector_loop: Send echo to node 10.228.191.13, status = failed: seastar::rpc::closed_ ERROR (connection is closed)
July 19th 2022, 10:53:56.328 ERROR 2022-07-19 10:53:56,328 [shard 2] rpc - client 10.228.191.13:7000: fail to connect: Connection refused
July 19th 2022, 10:53:54.241 WARN  2022-07-19 10:53:54,240 [shard 2] gossip - failure_detector_loop: Send echo to node 10.228.191.13, status = failed: seastar::rpc::unknown_verb_ ERROR (unknown verb)
July 19th 2022, 10:53:00.353 INFO  2022-07-19 10:53:00,352 [shard 0] storage_service - replace[a086fb9f-9d00-497b-86a9-e99c7a81aac6]: Removed replacing_node=10.228.191.13 to replace existing_node=10.228.198.13, coordinator=10.228.191.13
July 19th 2022, 10:53:00.353 INFO  2022-07-19 10:53:00,352 [shard 0] token_metadata - Removed node 10.228.191.13 as pending replacing endpoint which replaces existing node 10.228.198.13
July 19th 2022, 10:50:35.503 INFO  2022-07-19 10:50:35,502 [shard 0] token_metadata - Added node 10.228.191.13 as pending replacing endpoint which replaces existing node 10.228.198.13
July 19th 2022, 10:50:35.503 INFO  2022-07-19 10:50:35,502 [shard 0] storage_service - replace[a086fb9f-9d00-497b-86a9-e99c7a81aac6]: Added replacing_node=10.228.191.13 to replace existing_node=10.228.198.13, coordinator=10.228.191.13
July 19th 2022, 10:49:14.747 INFO  2022-07-19 10:49:14,746 [shard 0] gossip - InetAddress 10.228.191.13 is now UP, status = UNKNOWN
July 19th 2022, 10:49:13.955 INFO  2022-07-19 10:49:13,954 [shard 0] gossip - 60000 ms elapsed, 10.228.191.13 gossip quarantine over
July 19th 2022, 10:48:13.602 INFO  2022-07-19 10:48:13,602 [shard 0] gossip - FatClient 10.228.191.13 has been silent for 30000ms, removing from gossip
...
```

Decoded stacktrace:
```
decode stacktrace:
/opt/scylladb/libexec/scylla
/opt/scylladb/libexec/scylla: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 3.2.0, BuildID[sha1]=9c658935c77231a9f3eccfe202ae5f177914a333, stripped

[Backtrace #0]
seastar::noncopyable_function<void ()>::operator()() const at ./build/release/seastar/./seastar/include/seastar/util/noncopyable_function.hh:209
 (inlined by) seastar::thread_context::main() at ./build/release/seastar/./seastar/src/core/thread.cc:299
operator() at ./gms/gossiper.cc:1709
 (inlined by) ~deferred_action at ././seastar/include/seastar/util/defer.hh:63
 (inlined by) gms::gossiper::apply_new_states(gms::inet_address, gms::endpoint_state&, gms::endpoint_state const&) at ./gms/gossiper.cc:1741
/opt/scylladb/libreloc/libstdc++.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, BuildID[sha1]=bdd633ff5da0bba64d19ecf277a9eed7001da127, stripped

__cxa_throw_bad_array_new_length at ??:?
/opt/scylladb/libreloc/libc.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked (uses shared libs), BuildID[sha1]=9de0f6e78ec16db4c6c167efc6b37466705edeff, for GNU/Linux 3.2.0, stripped

abort at ??:?
operator() at ./gms/gossiper.cc:650
 (inlined by) _ZSt13__invoke_implIvZZZZN3gms8gossiper19apply_state_locallyESt3mapINS0_12inet_addressENS0_14endpoint_stateESt4lessIS3_ESaISt4pairIKS3_S4_EEEENK3$_7clIRN5utils14chunked_vectorIS3_Lm131072EEERSB_EEDaOT_OT0_ENKUlSK_E_clIRS3_EEN7seastar6futureIvEESK_ENUlvE_clEvEUlvE_JEESJ_St14__invoke_otherSM_DpOT1_ at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:61
 (inlined by) _ZSt8__invokeIZZZZN3gms8gossiper19apply_state_locallyESt3mapINS0_12inet_addressENS0_14endpoint_stateESt4lessIS3_ESaISt4pairIKS3_S4_EEEENK3$_7clIRN5utils14chunked_vectorIS3_Lm131072EEERSB_EEDaOT_OT0_ENKUlSK_E_clIRS3_EEN7seastar6futureIvEESK_ENUlvE_clEvEUlvE_JEENSt15__invoke_resultISJ_JDpT0_EE4typeESK_DpOSW_ at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:96
 (inlined by) _ZSt12__apply_implIZZZZN3gms8gossiper19apply_state_locallyESt3mapINS0_12inet_addressENS0_14endpoint_stateESt4lessIS3_ESaISt4pairIKS3_S4_EEEENK3$_7clIRN5utils14chunked_vectorIS3_Lm131072EEERSB_EEDaOT_OT0_ENKUlSK_E_clIRS3_EEN7seastar6futureIvEESK_ENUlvE_clEvEUlvE_St5tupleIJEEJEEDcSK_SM_St16integer_sequenceImJXspT1_EEE at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/tuple:1843
 (inlined by) _ZSt5applyIZZZZN3gms8gossiper19apply_state_locallyESt3mapINS0_12inet_addressENS0_14endpoint_stateESt4lessIS3_ESaISt4pairIKS3_S4_EEEENK3$_7clIRN5utils14chunked_vectorIS3_Lm131072EEERSB_EEDaOT_OT0_ENKUlSK_E_clIRS3_EEN7seastar6futureIvEESK_ENUlvE_clEvEUlvE_St5tupleIJEEEDcSK_SM_ at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/tuple:1854
 (inlined by) _ZN7seastar8futurizeIvE5applyIZZZZN3gms8gossiper19apply_state_locallyESt3mapINS3_12inet_addressENS3_14endpoint_stateESt4lessIS6_ESaISt4pairIKS6_S7_EEEENK3$_7clIRN5utils14chunked_vectorIS6_Lm131072EEERSE_EEDaOT_OT0_ENKUlSN_E_clIRS6_EENS_6futureIvEESN_ENUlvE_clEvEUlvE_JEEESU_SN_OSt5tupleIJDpT0_EE at ././seastar/include/seastar/core/future.hh:2099
 (inlined by) operator() at ././seastar/include/seastar/core/thread.hh:258
 (inlined by) _ZN7seastar20noncopyable_functionIFvvEE17direct_vtable_forIZNS_5asyncIZZZZN3gms8gossiper19apply_state_locallyESt3mapINS5_12inet_addressENS5_14endpoint_stateESt4lessIS8_ESaISt4pairIKS8_S9_EEEENK3$_7clIRN5utils14chunked_vectorIS8_Lm131072EEERSG_EEDaOT_OT0_ENKUlSP_E_clIRS8_EENS_6futureIvEESP_ENUlvE_clEvEUlvE_JEEENS_8futurizeINSt9result_ofIFNSt5decayISO_E4typeEDpNS11_IT0_E4typeEEE4typeEE4typeENS_17thread_attributesESP_DpOS14_EUlvE_E4callEPKS2_ at ././seastar/include/seastar/util/noncopyable_function.hh:124
std::terminate() at ??:?
?? ??:0
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at ./build/release/seastar/./seastar/include/seastar/util/backtrace.hh:59
 (inlined by) seastar::backtrace_buffer::append_backtrace() at ./build/release/seastar/./seastar/src/core/reactor.cc:756
 (inlined by) seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:786
seastar::print_with_backtrace(char const*, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:798
 (inlined by) seastar::sigabrt_action() at ./build/release/seastar/./seastar/src/core/reactor.cc:3716
 (inlined by) operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:3698
 (inlined by) __invoke at ./build/release/seastar/./seastar/src/core/reactor.cc:3694
?? ??:0
std::rethrow_exception(std::__exception_ptr::exception_ptr) at ??:?
gsignal at ??:?
gms::gossiper::do_apply_state_locally(gms::inet_address, gms::endpoint_state const&, bool) at ./gms/gossiper.cc:588
```

We have a patch (nearly identical to https://github.com/scylladb/scylla/pull/11081/files) that changes the interval to 300s.
We are wondering if it can be related to the quarantine delay set to 30s: https://github.com/scylladb/scylla/blob/6cdd5b9ff543755442ff680fbdebfa866cd4f856/gms/gossiper.cc#L68

What do you think @asias ?

**Comment by eliransin** [comment]
@asias I see all patches only refs this issue. What should be done to get it really fixed (or maybe we don't know) ?

**Comment by asias** [comment]
> @asias I see all patches only refs this issue. What should be done to get it really fixed (or maybe we don't know) ?

I am using refs instead of fixes because we need multiple patches to improve the problem. All the patches are now merged. I think the issue should be solved. 

I will request the backport once it stays in master for some time to get some more testing.

**Comment by asias** [comment]
> We had a crash of the whole cluster at the end of a replace operation:
> 
> Here is the log of a node that crashed (new node is `10.228.191.13`):
> 
> ```
> July 19th 2022, 23:47:01.448  0x40aa931
> July 19th 2022, 23:47:01.448 Aborting on shard 0.
> July 19th 2022, 23:47:01.448  0x299842c
> July 19th 2022, 23:47:01.448  /opt/scylladb/libreloc/libstdc++.so.6+0xad306
> July 19th 2022, 23:47:01.448  /opt/scylladb/libreloc/libc.so.6+0x268a3
> July 19th 2022, 23:47:01.448  0x29dafba
> July 19th 2022, 23:47:01.448  /opt/scylladb/libreloc/libc.so.6+0x3d2a1
> July 19th 2022, 23:47:01.448  0x7fd78d8e7a1f
> July 19th 2022, 23:47:01.448 terminate called after throwing an instance of 'std::runtime_ ERROR'
> July 19th 2022, 23:47:01.448  0x29a0050
> July 19th 2022, 23:47:01.448  0x3e27018
> July 19th 2022, 23:47:01.448  what():  Host 10.228.191.13 does not have HOST_ID application_state
> July 19th 2022, 23:47:01.448  0x3e57e72
> July 19th 2022, 23:47:01.448  /opt/scylladb/libreloc/libstdc++.so.6+0xad29b
> July 19th 2022, 23:47:01.448  0xf4e6fa
> July 19th 2022, 23:47:01.448  /opt/scylladb/libreloc/libstdc++.so.6+0xa1a45
> July 19th 2022, 23:47:01.448 Backtrace:
> July 19th 2022, 23:46:44.077 INFO  2022-07-19 23:46:44,077 [shard 0] storage_service - replace[49272acc-b8f9-4431-97c0-a629c58722e5]: Marked ops done from coordinator=10.228.191.13
> ... (elapsed to beginning of bootstrap, grep on "10.228.191.13")
> July 19th 2022, 23:46:44.077 INFO  2022-07-19 23:46:44,077 [shard 0] storage_service - replace[49272acc-b8f9-4431-97c0-a629c58722e5]: Marked ops done from coordinator=10.228.191.13
> July 19th 2022, 11:07:57.203 WARN  2022-07-19 11:07:57,203 [shard 0] gossip - Node 10.228.191.13 does not contain SUPPORTED_FEATURES in gossip or system table
> July 19th 2022, 11:06:10.367 INFO  2022-07-19 11:06:10,366 [shard 0] token_metadata - Added node 10.228.191.13 as pending replacing endpoint which replaces existing node 10.228.198.13
> July 19th 2022, 11:06:10.367 INFO  2022-07-19 11:06:10,366 [shard 0] storage_service - replace[49272acc-b8f9-4431-97c0-a629c58722e5]: Added replacing_node=10.228.191.13 to replace existing_node=10.228.198.13, coordinator=10.228.191.13
> July 19th 2022, 11:05:42.534 WARN  2022-07-19 11:05:42,534 [shard 0] gossip - Node 10.228.191.13 does not contain SUPPORTED_FEATURES in gossip or system table
> July 19th 2022, 11:04:51.655 WARN  2022-07-19 11:04:51,654 [shard 0] gossip - Node 10.228.191.13 does not contain SUPPORTED_FEATURES in gossip or system table
> July 19th 2022, 11:04:51.516 INFO  2022-07-19 11:04:51,515 [shard 0] gossip - InetAddress 10.228.191.13 is now UP, status = UNKNOWN
> July 19th 2022, 11:04:51.328 INFO  2022-07-19 11:04:51,328 [shard 0] gossip - 60000 ms elapsed, 10.228.191.13 gossip quarantine over
> July 19th 2022, 11:03:50.991 INFO  2022-07-19 11:03:50,991 [shard 0] gossip - FatClient 10.228.191.13 has been silent for 30000ms, removing from gossip
> July 19th 2022, 11:03:20.108 WARN  2022-07-19 11:03:20,108 [shard 0] gossip - Node 10.228.191.13 does not contain SUPPORTED_FEATURES in gossip or system table
> July 19th 2022, 11:03:20.041 INFO  2022-07-19 11:03:20,040 [shard 0] gossip - InetAddress 10.228.191.13 is now UP, status = UNKNOWN
> July 19th 2022, 10:56:18.434 INFO  2022-07-19 10:56:18,433 [shard 0] gossip - 60000 ms elapsed, 10.228.191.13 gossip quarantine over
> July 19th 2022, 10:55:18.124 INFO  2022-07-19 10:55:18,124 [shard 0] gossip - FatClient 10.228.191.13 has been silent for 30000ms, removing from gossip
> July 19th 2022, 10:53:56.328 WARN  2022-07-19 10:53:56,328 [shard 2] gossip - failure_detector_loop: Send echo to node 10.228.191.13, status = failed: seastar::rpc::closed_ ERROR (connection is closed)
> July 19th 2022, 10:53:56.328 ERROR 2022-07-19 10:53:56,328 [shard 2] rpc - client 10.228.191.13:7000: fail to connect: Connection refused
> July 19th 2022, 10:53:54.241 WARN  2022-07-19 10:53:54,240 [shard 2] gossip - failure_detector_loop: Send echo to node 10.228.191.13, status = failed: seastar::rpc::unknown_verb_ ERROR (unknown verb)
> July 19th 2022, 10:53:00.353 INFO  2022-07-19 10:53:00,352 [shard 0] storage_service - replace[a086fb9f-9d00-497b-86a9-e99c7a81aac6]: Removed replacing_node=10.228.191.13 to replace existing_node=10.228.198.13, coordinator=10.228.191.13
> July 19th 2022, 10:53:00.353 INFO  2022-07-19 10:53:00,352 [shard 0] token_metadata - Removed node 10.228.191.13 as pending replacing endpoint which replaces existing node 10.228.198.13
> July 19th 2022, 10:50:35.503 INFO  2022-07-19 10:50:35,502 [shard 0] token_metadata - Added node 10.228.191.13 as pending replacing endpoint which replaces existing node 10.228.198.13
> July 19th 2022, 10:50:35.503 INFO  2022-07-19 10:50:35,502 [shard 0] storage_service - replace[a086fb9f-9d00-497b-86a9-e99c7a81aac6]: Added replacing_node=10.228.191.13 to replace existing_node=10.228.198.13, coordinator=10.228.191.13
> July 19th 2022, 10:49:14.747 INFO  2022-07-19 10:49:14,746 [shard 0] gossip - InetAddress 10.228.191.13 is now UP, status = UNKNOWN
> July 19th 2022, 10:49:13.955 INFO  2022-07-19 10:49:13,954 [shard 0] gossip - 60000 ms elapsed, 10.228.191.13 gossip quarantine over
> July 19th 2022, 10:48:13.602 INFO  2022-07-19 10:48:13,602 [shard 0] gossip - FatClient 10.228.191.13 has been silent for 30000ms, removing from gossip
> ...
> ```
> 
> Decoded stacktrace:
> 
> ```
> decode stacktrace:
> /opt/scylladb/libexec/scylla
> /opt/scylladb/libexec/scylla: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 3.2.0, BuildID[sha1]=9c658935c77231a9f3eccfe202ae5f177914a333, stripped
> 
> [Backtrace #0]
> seastar::noncopyable_function<void ()>::operator()() const at ./build/release/seastar/./seastar/include/seastar/util/noncopyable_function.hh:209
>  (inlined by) seastar::thread_context::main() at ./build/release/seastar/./seastar/src/core/thread.cc:299
> operator() at ./gms/gossiper.cc:1709
>  (inlined by) ~deferred_action at ././seastar/include/seastar/util/defer.hh:63
>  (inlined by) gms::gossiper::apply_new_states(gms::inet_address, gms::endpoint_state&, gms::endpoint_state const&) at ./gms/gossiper.cc:1741
> /opt/scylladb/libreloc/libstdc++.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, BuildID[sha1]=bdd633ff5da0bba64d19ecf277a9eed7001da127, stripped
> 
> __cxa_throw_bad_array_new_length at ??:?
> /opt/scylladb/libreloc/libc.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked (uses shared libs), BuildID[sha1]=9de0f6e78ec16db4c6c167efc6b37466705edeff, for GNU/Linux 3.2.0, stripped
> 
> abort at ??:?
> operator() at ./gms/gossiper.cc:650
>  (inlined by) _ZSt13__invoke_implIvZZZZN3gms8gossiper19apply_state_locallyESt3mapINS0_12inet_addressENS0_14endpoint_stateESt4lessIS3_ESaISt4pairIKS3_S4_EEEENK3$_7clIRN5utils14chunked_vectorIS3_Lm131072EEERSB_EEDaOT_OT0_ENKUlSK_E_clIRS3_EEN7seastar6futureIvEESK_ENUlvE_clEvEUlvE_JEESJ_St14__invoke_otherSM_DpOT1_ at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:61
>  (inlined by) _ZSt8__invokeIZZZZN3gms8gossiper19apply_state_locallyESt3mapINS0_12inet_addressENS0_14endpoint_stateESt4lessIS3_ESaISt4pairIKS3_S4_EEEENK3$_7clIRN5utils14chunked_vectorIS3_Lm131072EEERSB_EEDaOT_OT0_ENKUlSK_E_clIRS3_EEN7seastar6futureIvEESK_ENUlvE_clEvEUlvE_JEENSt15__invoke_resultISJ_JDpT0_EE4typeESK_DpOSW_ at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:96
>  (inlined by) _ZSt12__apply_implIZZZZN3gms8gossiper19apply_state_locallyESt3mapINS0_12inet_addressENS0_14endpoint_stateESt4lessIS3_ESaISt4pairIKS3_S4_EEEENK3$_7clIRN5utils14chunked_vectorIS3_Lm131072EEERSB_EEDaOT_OT0_ENKUlSK_E_clIRS3_EEN7seastar6futureIvEESK_ENUlvE_clEvEUlvE_St5tupleIJEEJEEDcSK_SM_St16integer_sequenceImJXspT1_EEE at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/tuple:1843
>  (inlined by) _ZSt5applyIZZZZN3gms8gossiper19apply_state_locallyESt3mapINS0_12inet_addressENS0_14endpoint_stateESt4lessIS3_ESaISt4pairIKS3_S4_EEEENK3$_7clIRN5utils14chunked_vectorIS3_Lm131072EEERSB_EEDaOT_OT0_ENKUlSK_E_clIRS3_EEN7seastar6futureIvEESK_ENUlvE_clEvEUlvE_St5tupleIJEEEDcSK_SM_ at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/tuple:1854
>  (inlined by) _ZN7seastar8futurizeIvE5applyIZZZZN3gms8gossiper19apply_state_locallyESt3mapINS3_12inet_addressENS3_14endpoint_stateESt4lessIS6_ESaISt4pairIKS6_S7_EEEENK3$_7clIRN5utils14chunked_vectorIS6_Lm131072EEERSE_EEDaOT_OT0_ENKUlSN_E_clIRS6_EENS_6futureIvEESN_ENUlvE_clEvEUlvE_JEEESU_SN_OSt5tupleIJDpT0_EE at ././seastar/include/seastar/core/future.hh:2099
>  (inlined by) operator() at ././seastar/include/seastar/core/thread.hh:258
>  (inlined by) _ZN7seastar20noncopyable_functionIFvvEE17direct_vtable_forIZNS_5asyncIZZZZN3gms8gossiper19apply_state_locallyESt3mapINS5_12inet_addressENS5_14endpoint_stateESt4lessIS8_ESaISt4pairIKS8_S9_EEEENK3$_7clIRN5utils14chunked_vectorIS8_Lm131072EEERSG_EEDaOT_OT0_ENKUlSP_E_clIRS8_EENS_6futureIvEESP_ENUlvE_clEvEUlvE_JEEENS_8futurizeINSt9result_ofIFNSt5decayISO_E4typeEDpNS11_IT0_E4typeEEE4typeEE4typeENS_17thread_attributesESP_DpOS14_EUlvE_E4callEPKS2_ at ././seastar/include/seastar/util/noncopyable_function.hh:124
> std::terminate() at ??:?
> ?? ??:0
> void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at ./build/release/seastar/./seastar/include/seastar/util/backtrace.hh:59
>  (inlined by) seastar::backtrace_buffer::append_backtrace() at ./build/release/seastar/./seastar/src/core/reactor.cc:756
>  (inlined by) seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:786
> seastar::print_with_backtrace(char const*, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:798
>  (inlined by) seastar::sigabrt_action() at ./build/release/seastar/./seastar/src/core/reactor.cc:3716
>  (inlined by) operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:3698
>  (inlined by) __invoke at ./build/release/seastar/./seastar/src/core/reactor.cc:3694
> ?? ??:0
> std::rethrow_exception(std::__exception_ptr::exception_ptr) at ??:?
> gsignal at ??:?
> gms::gossiper::do_apply_state_locally(gms::inet_address, gms::endpoint_state const&, bool) at ./gms/gossiper.cc:588
> ```
> 
> We have a patch (nearly identical to https://github.com/scylladb/scylla/pull/11081/files) that changes the interval to 300s. We are wondering if it can be related to the quarantine delay set to 30s:
> 
> https://github.com/scylladb/scylla/blob/6cdd5b9ff543755442ff680fbdebfa866cd4f856/gms/gossiper.cc#L68-L72
> 
> What do you think @asias ?

Hello @geobeau 

First, I want to know if the cluster goes back to normal now?

The crash is similar to https://github.com/scylladb/scylla/pull/9658

The replacing node was removed by other nodes in gossip while the replacing node published a state to be applied  which needs host id.


As a workaround, you can increase the ring_delay to a  bigger number.  

Can you share the full log for 10.228.191.13 (replacing node) and one of existing node? Setting the dealy to 30s is not  a problem. A smaller  number will make the problem happen more easily.

**Comment by geobeau** [comment]
> First, I want to know if the cluster goes back to normal now? 

Yes, it went back and the issue didn't reproduce. I set the ring_delay to 300s since. Unfortunately, I don't have anymore the logs of incident.

**Comment by asias** [comment]
> > First, I want to know if the cluster goes back to normal now?
> 
> Yes, it went back and the issue didn't reproduce. I set the ring_delay to 300s since. Unfortunately, I don't have anymore the logs of incident.

OK. Good to hear it is back to normal.

**Comment by geobeau** [comment]
Hello @asias, with @pgoron we backported your patches (10x + speed up gossip) and we still have long settle time. After investigating deeper it seems to be due to `calculate_effective_replication_map` that is very inefficient with `everywhere_replication_strategy`. 

The path is something like:
`handle_state_normal` -> `replicate_to_all_cores` -> `create_effective_replication_map` -> `calculate_effective_replication_map`

https://github.com/scylladb/scylladb/blob/be44fd63f9ab8eb964eab91bf2202bf9e881a720/locator/abstract_replication_strategy.cc#L309-L319
Here the for loop is calling 17920 (256 * 70) to `calculate_natural_endpoints`.

On our cluster, it takes ~15 seconds to execute it and it's an issue only for `everywhere_replication_strategy`.
It echoes your fix: https://github.com/scylladb/scylladb/pull/10850/commits/6152f5b858d0cbd9545ac5558128fd283ddf3d4b but it's on another code path

**Comment by bhalevy** [comment]
Not only that, `everywhere_replication_strategy::calculate_natural_endpoints` doesn't yield.

https://github.com/scylladb/scylladb/blob/be44fd63f9ab8eb964eab91bf2202bf9e881a720/locator/everywhere_replication_strategy.cc#L22-L25

**Comment by geobeau** [comment]
I have pushed this commit on our fork: https://github.com/criteo-forks/scylla/commit/994c6ecf3c4edd151cd1db2ed0c92f32dc832cbb
It seems to have improved a lot the startup process

**Comment by npeshek-spotx** [comment]
@geobeau: Yep. That dropped our startup from 3000+ seconds to about 40.

**Comment by asias** [comment]
> I have pushed this commit on our fork: [criteo-forks@994c6ec](https://github.com/criteo-forks/scylla/commit/994c6ecf3c4edd151cd1db2ed0c92f32dc832cbb) It seems to have improved a lot the startup process. It's still a bit slow because we have the auth keyspace with high replicate number (RF=all for system_auth, see: #1796).

@geobeau Very nice! So, my  (10x + speed up gossip) + your https://github.com/criteo-forks/scylla/commit/994c6ecf3c4edd151cd1db2ed0c92f32dc832cbb, seem to speed up your cluster a lot. Thanks for testing it.

@geobeau Mind to send your fix as a PR (your credit)?

**Comment by geobeau** [comment]
> Mind to send your fix as a PR (your credit)?

@bhalevy already made a better patch: https://github.com/scylladb/scylladb/pull/11277

On my side, it still takes ~600 gossip poll to settle because of the `system_auth` keyspace. We configured it with network topology and RF=70 (= nodes in clusters). Do you think there is room for optimization for this use case? The alternative for us is to move to `distributed everywhere topology`, do you think it's a good idea?

**Comment by avikivity** [comment]
We plan to move auth to everywhere (and later raft) too.

**Comment by avikivity** [comment]
Just check if it's using QUORUM now, it can be bad with RF=70.

**Comment by geobeau** [comment]
Switching `system_auth` to `distributed everywhere topology` did the trick, the gossip settling takes only maximum 10 to 20 seconds

**Comment by avikivity** [comment]
@bhalevy should we backport this? I think this is a regression in its way, introduced by effective_replication_map.

**Comment by avikivity** [comment]
@bhalevy pings

**Comment by bhalevy** [comment]
> @bhalevy pings

@avikivity what exactly do you want to backport?
There's #12733, but also #10850 and #10822

I think we can reduce the risk by backporting 482ee369d03dbb1b67d1d76692d3460be3be80a8 and ba919aa88aa462b91520f73d3b7c2e775caaee87 instead, to prevent aborting the node operation on peers, even if the _prepare phase took longer than needed.

Ref #11011 #11078 #12969

Cc @asias @tgrabiec

**Comment by avikivity** [comment]
@bhalevy I was asking you, since I don't know what we want to backport.

**Comment by bhalevy** [comment]
There were quite a few changes related to `abstract_replication_strategy::get_address_ranges` and in particular with everywhere replication strategy.  I can prepare backport branched to trickle down the fixes top-down from branch-5.2 backwards toward branch-5.0, but I'm concerned the risk may outweigh the benefits and that we may stop short at branch-5.2 to get the comprehensive list of fixes, while for older branches, stick with https://github.com/scylladb/scylladb/commit/482ee369d03dbb1b67d1d76692d3460be3be80a8 and https://github.com/scylladb/scylladb/commit/ba919aa88aa462b91520f73d3b7c2e775caaee87 as a workaround and continue to live with reactor stalls in node operations with large clusters (but the node operation should be less susceptible to failures (unless the stalls will reach the increased node ops watchdog interval of 120 seconds).

**Comment by asias** [comment]
#12980 fixed the issue in another perspective: allow the slowness of the get_address_ranges and friends (allow the prepare verb took longer). It is much safer and easier to backport.
